[RFC 0148] Pipe operator (#148)
NixOS/rfcs: rfcs/0148-pipe-operator.md

---
feature: pipe-operator
start-date: 2023-05-23
author: @piegamesde
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Introduce a new "pipe" operator, `|>`, to the Nix language, defined as `f a` = `a |> f`.
Additionally, elevate `lib.pipe` to a built-in function.

As a reminder, `pipe a [ f g h ]` is defined as `h (g (f a))`.

# Motivation
[motivation]: #motivation

Creating advanced data processing like transforming a list is a thing commonly done in nixpkgs.
Yet the language has no support for function concatentation/composition,
which results in such constructs looking unwieldy and difficult to format well.
`lib.pipe` may be the most powerful library function with that regard,
but it is unknown and overlooked by many because it is not easily discoverable:

> Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs

Reviewer: I don't think I've even see `lib.pipe` anywhere. I'm guessing that it is the case for A LOT of people.

I think this argument says much more about the state of the documentation and the bizarre `builtins` vs `nixpkgs.lib` vs `nixos-lib` fenomena we have.

I think this would become a non-issue if `nixpkgs.lib` contained only functions that did something internal to Nixpkgs specifically (such as `lib.options`), and everything that manipulates data like strings, lists and attrsets was moved to `builtins` and folded nicely. By "folding" I mean that I think `map`, `mapAttrs`, `mapAttrs` flavors and `mapNullable` should exist in the same section in the manual. If I know how `map` works in the general sense, then I'll be able to understand how `mapAttrs` works. I would then be able to see all of the types `map` is applicable on and the different variations of the function by clicking on the "`map` function" header.

That said, this is outside of the scope of this RFC, so I digress. I do think that the issue lies in the documentation and technical structure though, and it would be nice if we started talking about that.

> Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: mostly irrelevant to this RFC.

(`rg '[\. ]pipe .* \['`).

> Additionally, it is not accessible to Nix code outside of nixpkgs,

Reviewer: Once again, this is a problem of `builtins` offering the bare minimum and `nixpkgs.lib` offerring a very rich library of generally-applicable functions even outside of Nixpkgs. This is a different problem.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: inclined to leave `pipe` in `lib`.

and due to Nix's lazy evaluation debugging type errors is really difficult.

Let's have a look at an arbitrarily chosen snippet of Nixpkgs code:

```nix
defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
  // ${value.reason}
  pref("${key}", ${builtins.toJSON value.value});
'') defaultPrefs));
```

It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:

```nix
defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
  lib.concatStringsSep "\n" (
    lib.mapAttrsToList
    (
      key: value: ''
        // ${value.reason}
        pref("${key}", ${builtins.toJSON value.value});
      ''
    )
    defaultPrefs
  )
);
```

One can observe the following issues:

- If you want to follow the data flow, you must read it from bottom to top,

>   from the inside to the outside (the input here is `defaultPrefs`).

Reviewer 3: Conversely, if you want to follow the evaluation order, you must read it in its natural order, from top to bottom.

This negates the point for me.

Note that recent Nix releases (2.13?) also show error traces in evaluation order: the last steps are printed last.

Mini explanation:

```
nix-repl> f = builtins.trace "f" (x: builtins.trace "f body" x) 

nix-repl> g = builtins.trace "g" (y: builtins.trace "g body" y) 

nix-repl> f (g null)
trace: f
trace: f body
trace: g
trace: g body
null
```

Author: Most of the time when writing Nix code, I don't really care about the specific order in which Nix evaluates things. I care a lot more about the intention and the order of the steps that process my data.

That traces show in evaluation order is not a contradiction, because traces are mostly used for debugging, which is also one of the rare situations where I do care about evaluation order.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: there is value to a `|>` operator, inclined to leave `pipe` in `lib`.

- Adding a function call to the output would require wrapping the entire
  expression in parentheses and increasing its indentation.

Compare this to the equivalent call with `lib.pipe`:

```nix
defaultPrefsFile = pipe defaultPrefs [
  (lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  ))
  (lib.concatStringsSep "\n")
  (pkgs.writeText "nixos-default-prefs.js")

> ];

Reviewer 3: This is not an improvement, because now we have to look up the order in which `pipe` applies the functions. The parentheses made that explicit, and they were easy to understand, especially for casual Nix users who don't write Nix code every month to keep such details fresh in memory.

Reviewer 4: I think this criticism mostly has to do with the obscurity of `pipe` and not the actual flow. I only learned of `pipe` here, and if I had known about it before, I would probably have used it quite often.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: inclined to leave `pipe` in `lib`.

```

> The code now clearly reads from top to bottom in the order the data is processed,

Reviewer: While I agree that the code reads clearly, I think the "top to bottom" part is somewhat of a detriment, actually. One of the first things that a beginner learns about Nixlang is how the order of statements DOES NOT matter. Of course, you *should* write code from top to bottom, but it is not always possible or feasible, and Nixlang itself doesn't care about the ordering.

In practice, this means that a lot of the time while reading Nixlang code, you have to make mental notes about the stuff you haven't yet seen the definition or usage of, or jump around the code in the `goto`-like fashion. `pipe` being one of the exceptions where you MUST read from top to bottom is actually unintuitive here.

Author: > One of the first things that a beginner learns about Nixlang is how the order of statements DOES NOT matter.

Of course the order of statements does matter, what you mean here is that due to Nix being a pure lazy functional language the *execution* order of statements is not strictly dictated by their appearance order in code. Therefore it is even more important to focus on the ordering relation that actually does matter in practice, and for me that is data flow.

> In practice, this means that a lot of the time while reading Nixlang code, you have to make mental notes about the stuff you haven't yet seen the definition or usage of, or jump around the code in the goto-like fashion. pipe being one of the exceptions where you MUST read from top to bottom is actually unintuitive here.

I find this statement weird, because it could equally well also apply to any function application itself: "function application being one of the exceptions where you MUST read from the inside to the outside is actually unintuitive here"

If you care about the direction, there is always the possibility to introduce `<|` instead, which is equivalent to basic function application.

it is easy to add and remove processing steps at any point.

With a dedicated pipe operator, it would look like this:

```nix
defaultPrefsFile = defaultPrefs
  |> lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  )
  |> lib.concatStringsSep "\n"

>   |> pkgs.writeText "nixos-default-prefs.js";

Reviewer 3: At least the operator points in the direction of the data flow, which should help with remembering the meaning.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: there is value to a `|>` operator.

```

The artificial distinction between the first input and the functions via the list now is gone,
and so are the parentheses around the functions.

> With the lower syntax overhead, using the operator becomes attractive in more situations,

Reviewer 3: Syntax overhead is a rather vague term. Arguably any addition to the syntax can only add overhead?

Author: That was not what I meant. I meant the character overhead of the function: due to having to type `pipe` and two additional brackets, using this is only worth it for longer function chains.

whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
Having a dedicated operator also increases visibility and discoverability of the feature.

# Detailed design
[design]: #detailed-design

## `|>` operator

A new operator `|>` is introduced into the Nix language.

> Semantically, it is defined as the reverse of function application: `f a` = `a |> f`.

Reviewer 3: Semantically it _is_ function application. It is only syntactically reversed.

It is left-associative and has a binding strength one weaker than function application:
`a |> f |> g b |> h` = `h ((g b) (f a))`.

## `builtins.pipe`

> `lib.pipe`'s functionality is implemented as a built-in function.

Reviewer 3: `pipe` is not a good function because the order is not obvious. It's already available practically everywhere as `lib.pipe`. It's also hardly ever used, so doing this for performance strips it of the last argument for making it a builtin.

Builtins need to satisfy more stringent requirements, they take resources from the Nix team, and they can never be changed or removed.

I would recommend to always use a library, which _can_ change things arbitrarily, thanks to being locked or pinned.

Reviewer: > It's also hardly ever used

I think this is due to documentation issues rather than due to this function itself being bad. While it has a lot of potential of misuse, it can actually make code more readable if you put in the effort for your code to be readable from top to bottom.

> I would recommend to always use a library, which can change things arbitrarily, thanks to being locked or pinned

This... makes sense, but then I'd make an argument that parts of `nixpkgs.lib` that manipulate generic data should exist outside of `builtins` and outside of `nixpkgs` - in a separate repository. It is not weird: we already have library flakes like `flake-utils` or `flake-parts`.

This discussion is outside of the scrope of this RFC, though; I merely disagree that `lib.pipe` not seeing much usage is due to technical reasons.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: inclined to leave `pipe` in `lib`.

Reviewer 5: > pipe is not a good function because the order is not obvious.

@roberth for what it’s worth, the commit where I introduced it has a rationale for the order in its message.

Reviewer 3: It's not a problem with your contribution, but with the constraints of the Nix syntax.

`lib.pipe` is a necessary step, and its adoption shows that it is useful to many, so thank you for creating it!

The main motivation for this is that it allows to give better error messages
like line numbers when some part of the pipeline fails.
Additionally, it allows easy usage outside of Nixpkgs and increases discoverability.

> While Nixpkgs is bounds to minimum Nix versions and thus `|>` won't be available until

Reviewer: I think this is a great argument against going through with this RFC.

Author: This would be an argument for never making any changes to improve the language at all, ever. Unless you see some shiny alternative language to which we'll realistically all migrate to within the next five years, any improvements to the Nix language are still expected to be beneficial.

I've actually thought about proposing this change for quite some time now, and the main thing that held me back was the idea is that Nix is a "lost cause" and that I should rather spending my energy towards some successor. But eventually, I realized that realistically, we'll be stuck with Nix for a long time to come, so better invest the energy to make it a bit more comfortable …

Also note that outside of Nixpkgs, the feature will be available almost immediately. There is a lot of Nix code outside of Nixpkgs, all these flakes, shells, and users' system configurations. Projects like home-manager, devenv, nur, etc. are all free to migrate to newer Nix versions at their own pace. We shouldn't ignore these.

since one may treat the operator as syntactic sugar for function application.
No fundamentally new semantics are introduced to the language.

# Prior art

Nickel has `|>` too, with the same name and semantics.

F# has `|>`, called "pipe-forward" operator, with the same semantics.
Additionally, it also has "pipe-backward" `<|` and `>>`/`<<` for forwards and backwards function composition.
`<|` is equivalent to function application, however its lower binding order allows removing parentheses:
`g (f a)` = `g <| f a`. All these operators have the same precedence and are left-associative.

Elm has the same operators as F#.

Haskell has the (backwards) function composition operator `.` in its prelude: `(g . f) a` = `g (f a)`.
It also has "reverse application" `&`, which is roughly equivalent to `|>`,
and `$`, which is function application again but right-associative and very weakly binding.
`.` binds stronger than both.

`|>` is definable as an infix function in several other programming languages,
and in even more languages as macro or higher-order function (including Nix, that's `lib.pipe`).

> # Alternatives

Reviewer 6: I'd like to propose an alternative that I already previously commented in https://github.com/NixOS/nix/issues/1845#issuecomment-1520798023 and https://github.com/NixOS/rfcs/pull/135#issuecomment-1557132711: To have an `apply` syntax like this:

```
apply f;
apply g;
x

# The same as
f (g x)
```

This is effectively the `$` operator (which is why I originally proposed it in https://github.com/NixOS/nix/issues/1845), but with more nixy syntax.

Advantages:
- More natural syntax, fitting better into the current Nix language, since there's already `with ...;` and `assert ...;`.
- This would be consistent with the new (experimental) CLI's [`--apply` flag](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-eval.html?highlight=--apply#options), which works in the same way.
- More verbose, making it clearer what the code does, and doesn't need a lot of explanation
- (weak) Doesn't add another operator, keeping the language simpler regarding operator precedence.

Disadvantages:
- More verbose, it takes longer to type
- The order is inverse from a pipe operator, making it harder to read in some cases.

Examples:
```
# Hypothetical `lib.guard`
apply lib.guard { ensure = x != 0; message = "Cannot divide by zero"; };
10 / x

apply builtins.trace "evaluating condition";
if condition then "foo" else "bar"

# Grouping and concatenating a list of strings by their 2-letter prefix:
apply mapAttrs (name: concatMapStringsSep " ");
apply groupBy (substring 0 2);
[ "foo" "bar" "baz" ]
```

Author: Adding to the disadvantages list:

- It would be yet another language feature that Nix reinvents in a custom way\*, instead of using constructs that users might recognize from other languages. (Or is there any prior work in other languages to this? It kind of reminds me of Rust function attributes and Python decorators a bit, but those are still significantly different)
- (big) Requires a new keyword in the language, which is a lot trickier w.r.t. backwards compatibility than introducing a new operator

I'll definitely add this to the alternatives list, but unless there is a way to solve the backwards compatibility issues without having to depend on the introduction of Nix language versioning, I remain skeptical.

\* To be fair some of these predate the equivalents seen elsewhere as it is a pretty old language, but for features introduced today "doing things differently" requires some pretty convincing arguments IMO.

Reviewer 6: Yeah the backwards compatibility thing is a problem, but it's a more general problem that needs to be fixed at some point. https://github.com/NixOS/rfcs/pull/137 is being worked on by @fricklerhandwerk and @yorickvp (not committed though), which should pave a way towards being able to make such changes.

Reviewer 7: > Or is there any prior work in other languages to this

`apply` tactic in theorem provers. But it's a rather ad-hoc kind of tactic, `refine` is more general and allows doing things like `refine func arg (λx => ?_)` which works like `use <- func(arg)` [use expression](https://gleam.run/news/v0.25-introducing-use-expressions/) in gleam.

Reviewer 8: I think semicolon `;` already makes language much more confusing because the first thought that goes into it is that it's a statement end, rather than a separation like `,` (comma) in most other languages.

I would not recommend going anything with this semicolon approach, as this would better be attained by `let ...; in ...` anyway:

> ```
> apply f;
> apply g;
> x
> 
> # The same as
> f (g x)
> ```
@infinisil This doesn't make much sense other than breaking the function call into multiple lines, what we rather wanted to achieve with pipe operator was to invert the syntax so `input -> output` is clearly readable, rather than `f ( g <- input )` which has to be read as "input goes into g goes into f" to understand the flow.

Reviewer 9: At least for me, this `apply` syntax is very confusing. My first thought when seeing `apply f` was, "The `apply` function is being called on `f`, but the result is not used anywhere. Does `apply` have side effects? That is strange because I though Nix was a purely functional language. I guess `apply` breaks the purity of Nix.". I would have no idea what was actually happening if your example did not include the equivalent in simple Nix syntax.

Reviewer 10: Agree that it fits into the language syntax. 

But tbh. the syntax:

`foo = apply g; x;` 

seems unintuitive unless you know that `assert` requires two semicolons. 

I like the idea of having a pipe operator, over this syntax, because it feels more natural and follows established conventions in other languages (as mentioned in the rfc).

Reviewer 3: > `assert` requires two semicolons.

Technically only one. The last one is for a binding or attribute, not for the `assert` expression.

`assert true; 1` is a valid expression, as you can see by entering it in the repl or into a file.

Kinda proves the point about `apply f; a` though.

Reviewer 2: Shepherds currently believe this conversation can be resolved by adding a summary of this alternative to this section.

[alternatives]: #alternatives

For each change this RFC proposes, there is always the trivial alternative of not doing it. See #drawbacks.

## More operators

> We could use the occasion and introduce more operators like those mentioned above.

Reviewer: I think introducing the pipe operator is hard enough as it is. While using the occasion to also merge other operators makes sense, I don't see the future where it happens, given the organizational difficulties and us not having any mechanism to do that without treewide breakages: we don't have language versions.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: inclined to leave `pipe` in `lib`.

Function composition is mostly interesting for the so-called "point-free" programming style,
where partially applied compositions of functions are preferred over the introduction of lambda terms.
However, Nix is not well suited for that programming style for various reasons,
nor would that point-free style have nearly as many applications in real-world Nixpkgs code.

F#'s reverse-pipe operator has a lot less use due to its left-associativity,
but a right-associative version of it more similar to Haskell's `$` might be an alternative:

```nix
defaultPrefsFile = 
  pkgs.writeText "nixos-default-prefs.js" <|
  lib.concatStringsSep "\n" <|
  lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  )
  defaultPrefs
  ;
```

> Adding both pipe directions raises questions about how these two interact when used together.

Reviewer 3: Interaction without parentheses could be forbidden, but this is a kind of thing we do yet in the language, so it has some small extra implementation cost that has to be repeated in all serious implementations (e.g. `nix`, `hnix`, `tvix`, linters that care)

Author: Are there any precedents in other languages for forbidding to mix certain operators like this? (On a syntax level, ignoring types) How would one formalize such a grammar?

Reviewer 11: Just tried this in Haskell:
```
ghci> import Flow
ghci> id <| 42 |> id

<interactive>:2:1: error:
    Precedence parsing error
        cannot mix ‘<|’ [infixr 0] and ‘|>’ [infixl 0] in the same infix expression

Author: Turns out Scala has a similar restriction too.

F# has the same binding strength for both, but this only works well because both are left-associative.
Haskell has `&` stronger than `$`, which is very sensible but unlikely to be intuitive to a new user.
Given that we want to call them `|>` and `<|` instead, then users might equally well to assume both have
equal strength.

Given these restrictions and the fact that situations where one needs both in Nix are expected to be fairly rare,
it is recommended to choose either one of `|>` and `<|`, but not have both in the language.

## Change the `pipe` function signature

There are many equivalent ways to declare this function, instead of just using the current design.

> For example, one could flip its arguments to allow a partially-applied point-free style (see above).

Reviewer 3: I think you could call this _curried style_.
While it is point free in the technical sense of not having a variable name to carry the data flow, point-free is generally only used in a context where higher order functions are used for more arbitrary data flows. For many, it also carries the connotation of the synonymously used _pointless style_.

Author: I've never heard that term and the resources I've read so far all talked about "point-free" programming style. Google does not seem to know much either, so some links would be appreciated.

Reviewer 5: point-free is also point-less, in a semantic sense :stuck_out_tongue:

Reviewer 2: Shepherds currently believe this conversation can be resolved by accepting or rejecting the suggestion below.

> One could also make this a single-argument function so that it only takes the list as argument.

Reviewer 3: This suggests that the list is heterogeneous: while most elements must be functions, one of them could be of any type (other elements permitting). Heterogeneous lists would present another unnecessary surprise to casual readers.

Author: Another downside of having everything in one list is that one loses the last hint of direction the `pipe` function currently has: by passing the argument first and then the functions, it makes a lot more sense to apply the following functions from top to bottom and not the other way around.

> For example, one could flip its arguments to allow a partially-applied point-free style (see above).
> One could also make this a single-argument function so that it only takes the list as argument.

Reviewer 3: Continuing the conversation on the previous line with a suggestion.

```suggestion
For example, one could flip its arguments so that it becomes function composition on list of functions.
Not only can this function trivially replace pipe; it can also be readily used where a function is expected, such as in `map`.
```

Reviewer 2: @piegamesde, :+1: or :-1:?

# Drawbacks
[drawbacks]: #drawbacks

- Introducing `|>` has the drawback of adding complexity to the language, and it will break older tooling.
- The main purpose of `builtins.pipe` is as a stop-gap until Nixpkgs can use `|>`. After that, it will be mostly redundant.

# Unresolved questions
[unresolved]: #unresolved-questions

- Who is going to implement this in Nix?
- How difficult will the implementation be?
- Will this affect evaluation performance in some way?
  - There is reason to expect that replacing `lib.pipe` with a builtin will reduce its overhead,
    and that the builtin should have little to no overhead compared to regular function application.

# Future work
[future]: #future-work

Once introduced and usable in Nixpkgs, existing code may benefit from being migrated to using these features.
Automatically transforming nested function calls into pipelines is unlikely,
as doing so is not guaranteed to always be a subjective improvement to the code.

> It might be possible to write a lint which detects opportunities for piping, for example in nixpkgs-hammering.

Reviewer 3: Overly zealous rules like this make people focus on the arbitrary rather than solving the actual problems.
You could fill your day bothering people with style suggestions, but that's not going to help anyone.
We have real unsolved problems in Nixpkgs and NixOS. How about getting out of your comfort zone and actually achieve something that solves a real packaging/deployment problem? (Not talking about this RFC necessarily, but about style rule policing)

Author: I'm not saying that we should do this, I'm just stating the possibility that we *could*.

---
feature: pipe-operator
start-date: 2023-05-23
author: @piegamesde

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)
> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 6: The more shepherds the merrier!

```suggestion
shepherd-team: @roberth @iFreilicht @illustris @adrian-gierakowski 
shepherd-leader: @iFreilicht 
```

@iFreilicht I hope you don't mind being leader, just means to make sure the RFC keeps moving forward, e.g. by scheduling meetings

Reviewer 12: Yes I'm cool with that :)

Reviewer 13: RFC Steering Committee accepts proposed nominations!

Reviewer 12: @piegamesde Can you merge the proposed change? Then we could maybe have schedule a kick-off and all get up to speed on what has to happen next.

Reviewer 6: Btw there might be some confusion: I already proposed this change on behalf of the steering committee originally. In today's meeting we just wanted to reconfirm that these are the shepherds.

      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  ))
  (lib.concatStringsSep "\n")
  (pkgs.writeText "nixos-default-prefs.js")
];
```

The code now clearly reads from top to bottom in the order the data is processed,
it is easy to add and remove processing steps at any point.

With a dedicated pipe operator, it would look like this:

```nix
defaultPrefsFile = defaultPrefs
  |> lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  )
  |> lib.concatStringsSep "\n"
  |> pkgs.writeText "nixos-default-prefs.js";
```

The artificial distinction between the first input and the functions via the list now is gone,
and so are the parentheses around the functions.
With the lower character overhead, using the operator becomes attractive in more situations,
whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
Having a dedicated operator also increases visibility and discoverability of the feature.

# Detailed design
[design]: #detailed-design

## `|>` operator

A new operator `|>` is introduced into the Nix language.

> It is defined as function application with the order of arguments swapped: `f a` = `a |> f`.

Author: Then I'd have to do that everywhere for consistency, and "=" expresses the same thing and is more concise.

> It is left-associative and has a binding strength one weaker than function application:

Reviewer 14: Currently we are parsing "function application" like this"

```
expr_op : '!' expr_op
        | '-' expr_op
        | expr_op BINARY_OP expr_op
        | expr_app

%right ->
%left ||
%left &&
%nonassoc == !=
%nonassoc < > <= >=
%right //
%left NOT
%left + -
%left * /
%right ++
%nonassoc '?'
%nonassoc NEGATE

expr_app : expr_app expr_select
         | expr_select
```

So what it's binding strength in comparison to other binary operators?

(i.e. how should we interpret this expression)

```
1 + 2 |> 3
```

Reviewer 3: IIUC, it'd be equivalent to `1 + (2 |> 3)`.

Without a type error, the example could be:
`1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`

Reviewer 14: > IIUC, it'd be equivalent to `1 + (2 |> 3)`.
> 
> Without a type error, the example could be: `1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`

Why not `(1 + 2) > 3` ? I think we should explicitly state the relationship in this RFC

Reviewer 15: > > IIUC, it'd be equivalent to `1 + (2 |> 3)`.
> > Without a type error, the example could be: `1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`
> 
> Why not `(1 + 2) > 3` ? I think we should explicitly state the relationship in this RFC. (i.e. the _precedence_ among other binary operators.)

100% agree.

I'm not a javascript person at all but they're spending a lot of time figuring out exactly this over at https://github.com/tc39/proposal-pipeline-operator . That proposal is for hack-style pipe operators, and they discuss the difference with so-called "F# style" pipe operators, which I think is what this here RFC 0148 is going for, right? See e.g. https://github.com/valtech-nyc/proposal-fsharp-pipelines?tab=readme-ov-file , particularly the links in that README. It could be worth picking a specific language (F#?) and imitating their syntax as closely as possible. This means we can reuse all their hard work in figuring out the ergonomics, and it makes it an easier sell to the community :) "Nix now supports F# pipes".

Reviewer 10: How about telling me where to insert "pipe" into this table. It specifies the precedence of all operators and should be updated accordingly once the pipe operator is introduced: https://nixos.org/manual/nix/stable/language/operators

Reviewer 10: @inclyc 

> Why not (1 + 2) > 3

The plus operator (+) has precedence of 7. While function application has precedence of 2.
If the `|>` operator had weaker strength than (+) this opens a whole new dimension of complexity. It would allow many binary/unary operations. (i.e., all those from 1 to 7). Giving function application and pipe operator neighboring precedence means they behave similarly which I think is good behavior since they do similar things.

Author: "a binding strength one weaker than function application." yeah seeing this in context, I think this is just wrong.

We also had the discussion about the binding strength and associativity of `|>` and `<|`, and came to the preliminary conclusion that it should be forbidden to mix both on the same level. Maybe we could expand this and restrict the expressions on either side of `|>` to not involve *any* other operators?

Reviewer 3: Also, forbidden constructs can be allowed in the future, whereas mistakenly allowed constructs require indefinite support unless we give up on long term reproducibility.

I would expect the pipe operators to form somewhat large expressions, where an extra pair of parentheses doesn't really matter much, and helps with clarity anyway.

So I agree that this is probably a good way to reduce scope and certainly reduce risk, although it *might* make the grammar more complicated and cause somewhat of a less obvious usage of some parser libraries/generators.

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: several options are viable here, and an experimental implementation is probably more useful than more talk for figuring out which is best.

> `a |> f |> g b |> h` = `h ((g b) (f a))`.

Reviewer 16: Explain this also in terms of `|>` + parenthesis - given that I genuinely understood nothing here

Indeed, as far as I remember, functions in Nix have only one argument (things like `a:b: a+b` are indeed a function receiving a function).
So, I'm confused about `h` here.

Reviewer 2: I find this line pretty straightforward.

Not sure if this helps, but imagine the variables here have the following types:
```haskell
a :: a
b :: b
f :: a -> t1
g :: b -> t1 -> t2
h :: t2 -> t3
```

So each element in the pipeline after the first is a function taking the previous type and returning the next one:
```haskell
f :: a -> t1
g b :: t1 -> t2
h :: t2 -> t3
```

Reviewer 2: Shepherds currently believe this conversation is resolved, outcome: RFC text does not need to change.

> ---

Reviewer 2: ### Plan for moving forward/stabilizing feature in Nix

Now that the experimental feature has been released, it may be time to start talking about a plan for evaluating it.

I see a number of possibilities:
* We do nothing and hope it happens organically that enough enthusiasts try the feature out to give it a good trial. In all likelihood, people will only comment if they dislike something about the feature, which means we could receive little or no feedback and be left wondering if enough people have looked at it. This requires least effort but is likely to take the longest.
* We promote the feature in official Nix channels and set up some sort of polling for both positive and negative feedback, with an expectation of how long the poll will be running before moving this RFC to FCP. This requires a bit of up-front work and may be biased towards people who pay attention to official Nix channels, but has the advantage of resolving faster than the first possibility.
* We devise an actual experiment to measure people's ability to learn the new operators. This might take the form of a survey asking people to rate the readability of various Nix expressions, some with the new operators and some without, and testing if there is a statistically significant difference in the ratings. We could send the survey to a representative sample of community members instead of (or in addition to) the public polling from the previous possibility, to try to control for bias. I'm not aware of the Nix community undertaking this level of rigor in the past but it would help us bring a stronger case for assuaging people's concerns about the mental cost of adding new operators. It would be a substantial investment of time from a few people (I'm happy to be one of those people but it probably shouldn't be me alone).

Anyone else have thoughts?

Reviewer 14: > We do nothing and hope it happens organically that enough enthusiasts try the feature out to give it a good trial. In all likelihood, people will only comment if they dislike something about the feature, which means we could receive little or no feedback and be left wondering if enough people have looked at it. This requires least effort but is likely to take the longest.

I think it would be beneficial to give the new feature more visibility and promotion, rather than leaving it as an experimental feature indefinitely.
Collecting feedback through some form of polling or voting mechanism could provide valuable insights.
Additionally, introducing new features in a programming language is a significant event, especially for a community-driven project like Nix. It's likely that experienced PL experts would be needed to assess its impact and stability.

> ---

Reviewer 2: (Attempting to corral feedback into a thread.)

@illustris

> ```nix
> nix-repl> double = x: x*2
> 
> # current syntax
> nix-repl> 3 |> (x: 1 + x) |> double |> toString
> "8"
> 
> # proposed alternative
> nix-repl> 3 |> x: 1 + x |> double |> toString
> 
> # current syntax
> nix-repl> 3 |> (x: 1 + (x |> double)) |> toString
> "7"
> 
> # proposed alternative
> nix-repl> 3 |> x: 1 + (x |> double) |> toString
> ```

Both of those proposals seem, to me, to be ‘more likely’ to mean something else:

```nix
3 |> (x: (1 + x |> double |> toString))
3 |> (x: (1 + (x |> double) |> toString))
```

In other words, if `|>` binds more loosely than the lambda-forming `:`, it would be the first thing in the language to do so and thus would be very surprising to me. So far in the language, a lambda always extends as far rightward as it can.

Author: ### Partially applied pipes

To me this smells like function composition with extra steps. Function composition is discussed in the RFC as well, and while I am not completely opposed to it, my fear is that it reduces code readability given a lack of type annotations. In `x = |> inc |> toString` it is a lot less intuitively clear that `x` is a function.

This is one of these questions where I'd like to see more usage to find out whether this is a thing that comes up sufficiently often in practice that it is worth dealing with.

(Side note: I think we would get this feature for free if Nix had operator sections)

### Lambdas inside pipe

While I agree that not requiring parentheses here would be nice, especially given that one goal of the pipe operator is to reduce parentheses, I don't see any way this would be realistically implementable without throwing the entire language under the bus. Making the binding strength of abstractions context dependent just doesn't sound great

Reviewer 5: > Making the binding strength of abstractions context dependent just doesn't sound great

agreed, it’s a recipe for disaster

Reviewer 17: ### Lambdas inside pipe

> Making the binding strength of abstractions context dependent

Right. I didn't think that through. This is a bad idea.

### Partially applied pipes

> In `x = |> inc |> toString` it is a lot less intuitively clear that `x` is a function

In my opinion it is slightly more clear than, for example,
```
concatLines = concatStringsSep "\n"
getBin = getOutput "bin"
```
or any other partially applied function. The open `|>` (or `|`) at the start suggests to me that this is a function.

feature: pipe-operator
start-date: 2023-05-23
author: @piegamesde
shepherd-team: @roberth @rhendric @illustris @adrian-gierakowski
shepherd-leader: @rhendric
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

> Introduce a new "pipe" operator, `|>`, to the Nix language, defined as `f a` = `a |> f`.

Reviewer 3: ```suggestion
Introduce new "pipe" operators to the Nix language, where
- `a |> f` is defined as `f a`, and similarly but flipped,
- `f <| a` is defined as `f a`
```

> Additionally, elevate `lib.pipe` to a built-in function.
> 
> As a reminder, `pipe a [ f g h ]` is defined as `h (g (f a))`.

Reviewer 3: ```suggestion
```

Neither Nix nor Lix seem to have implemented a `builtins.pipe` function.

```nix
defaultPrefsFile = pipe defaultPrefs [
  (lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  ))
  (lib.concatStringsSep "\n")
  (pkgs.writeText "nixos-default-prefs.js")
];
```

The code now clearly reads from top to bottom in the order the data is processed,
it is easy to add and remove processing steps at any point.

With a dedicated pipe operator, it would look like this:

```nix
defaultPrefsFile = defaultPrefs
  |> lib.mapAttrsToList (
    key: value: ''
      // ${value.reason}
      pref("${key}", ${builtins.toJSON value.value});
    ''
  )
  |> lib.concatStringsSep "\n"
  |> pkgs.writeText "nixos-default-prefs.js";
```

The artificial distinction between the first input and the functions via the list now is gone,
and so are the parentheses around the functions.
With the lower character overhead, using the operator becomes attractive in more situations,
whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
Having a dedicated operator also increases visibility and discoverability of the feature.

> 

Reviewer 3: ```suggestion

As a secondary effect, the addition of pipe operators reduces the incentive to add an unbounded number of semi-inconsistent helper functions to the Nixpkgs library. (i.e. ones that do not reach the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)).

```

# Detailed design
[design]: #detailed-design

## `|>` operator

A new operator `|>` is introduced into the Nix language.
It is defined as function application with the order of arguments swapped: `f a` = `a |> f`.
It is left-associative and has a binding strength weaker than function application:
`a |> f |> g b |> h` = `h ((g b) (f a))`.

> 

Reviewer 3: TODO:
- [ ] `<|` operator
  - can repurpose prose from alternatives section
- [ ] no mixing / associativity: none (in line with recent [piegames comment](https://github.com/NixOS/rfcs/pull/148#issuecomment-3626642721))
- [ ] precedence higher than all other current Nix operators, notably `->`. In other words, it is designed to divide "large" expressions - opposite of the usual (juxtaposition) function application operator.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

## Tooling support

Like any language extension, this will require the available Nix tooling to be updated.
Updating parsers should be pretty easy, as the syntax changes to the language are fairly minimal.
Tooling that evaluates Nix code in some way or does static code analysis should be easy to support too,
since one may treat the operator as syntactic sugar for function application.
No fundamentally new semantics are introduced to the language.

## Nixpkgs interaction

> As soon as the Nixpkgs minimum version contains `|>`, using it will be allowed and encouraged in the documentation.

Reviewer 18: Is it allowed to backport the pipe operator to the Nixpkgs minimum nix version?

Reviewer 19: Do major feature additions like this usually get a backport?

Even if it was backported, wouldn't it defeat the purpose of having a minimal version that's likely to be installed by most users?

A backport would still be a new release, even if it was based on an old major version number.

Reviewer 20: Sometimes.  Support for zstd cache compression was backported to 2.3.

The reason we have a minimum version of 2.3 not so much to support people who haven't updated Nix in several years, it's because later versions of Nix have [unresolved regressions](https://github.com/NixOS/nix/issues/9412) in things that some people depend on.

where partially applied compositions of functions are preferred over the introduction of lambda terms.
However, Nix is not well suited for that programming style for various reasons,
nor would that point-free style have nearly as many applications in typical Nixpkgs code.

Take for example this library function, written in a point-free style by using `flip pipe` as function concatenation operator:

```nix
concatMapAttrs = f: flip pipe [ (mapAttrs f) attrValues (foldl' mergeAttrs { }) ];
```

When reading this code, one has to manually do the headwork of inferring the types to understand what this function does.
In Haskell, its powerful type system and type inference would quickly spot any mistakes made.
But in Nix, this can lead to very confusing runtime errors instead
(even ignoring the additional stack trace noise of using `flip pipe`).
Compare this to the fully specifified version of the same function:

```nix
concatMapAttrs = f: v: pipe v [ (mapAttrs f) attrValues (foldl' mergeAttrs { }) ];
```

Would you have guessed correctly from the first code example whether it's `f: v:` or `v: f:`?

## Pipe-forward vs pipe-backward

> We could use `<|` instead of `|>` instead:

Reviewer 3: It seems we're adopting this

This could probably be fixed within Nixpkgs alone,
however not without incurring a significant performance penalty for using "reflection".
A built-in operator would be able to provide this more detailed error information basically for free.

Additionally, it allows easy usage outside of Nixpkgs and increases discoverability.

While Nixpkgs is bound to minimum Nix versions and thus `|>` won't be available until
several years after its initial implementation,
it can directly benefit from `builtins.pipe` and its better error diagnostic by overriding `lib.pipe`.
Elevating a Nixpkgs library function to a builtin has been done several times before,
for example `bitAnd`, `splitVersion` and `concatStringsSep`.

The main drawback is that once `|>` is available, there is little use for `builtins.pipe` anymore,
so the main purpose of that would be as a stop-gap for Nixpkgs
until the minimum Nix version is sufficiently high to allow using `|>`.

# Drawbacks
[drawbacks]: #drawbacks

- Introducing `|>` has the drawback of adding complexity to the language, and it will break older tooling.

# Unresolved questions
[unresolved]: #unresolved-questions

> - What is the precise binding strength of the operator?
> - Who is going to implement this in Nix?
> - How difficult will the implementation be?
> - Will this affect evaluation performance in some way?
>   - There is reason to expect that replacing `lib.pipe` with a builtin will reduce its overhead,
>     and that the builtin should have little to no overhead compared to regular function application.
> 
> In order to decide which operators to add to the language (see Alternatives),
> a larger survey across the Nixpkgs code will be conducted.
> This will give us quantitative information to better make any decisions involving tradeoffs.
> 

Reviewer 3: ```suggestion
```

I'm convinced by the qualitative case for these operators, and the other questions have either been answered explicitly or implicitly.
For instance the implementations have not suffered in performance by adding support, and I see no reason to expect problems from usage.
If anything, the pipe operator can be expected to be faster in AST-driven interpreters, because it dispatches directly into a function application, either just like `f a`, but faster than `lib.pipe` which involves multiple extra calls and a list construction.
